Src/HALAL/Services/Communication/Ethernet/TCP/ServerSocket.cpp

```diff
@@ -0,0 +1,198 @@
+		if(error == ERR_OK){
+			tcp_output(client_control_block);
+			tx_packet_buffer = tx_packet_buffer->next;
+			if(tx_packet_buffer != nullptr){
+				pbuf_ref(tx_packet_buffer);
+			}
+			memp_free_pool(memp_pools[PBUF_POOL_MEMORY_DESC_POSITION],temporal_packet_buffer);
+		}else if(error == ERR_MEM){			//Low on memory
+			tx_packet_buffer = temporal_packet_buffer;
+		}else{
+			//TODO: Error Handler
+		}
+	}
+}
+
+err_t ServerSocket::accept_callback(void* arg, struct tcp_pcb* incomming_control_block, err_t error){
+	tcp_setprio(incomming_control_block, priority);
+	priority++;
+	if(listening_sockets.contains(incomming_control_block->local_port)){
+		ServerSocket* server_socket = listening_sockets[incomming_control_block->local_port];
+		listening_sockets.erase(incomming_control_block->local_port);
+		tcp_nagle_disable(incomming_control_block);
+		tcp_arg(incomming_control_block, server_socket);
+		server_socket->state = ACCEPTED;
+		server_socket->client_control_block = incomming_control_block;
+		server_socket->tx_packet_buffer = nullptr;
+		tcp_recv(incomming_control_block, receive_callback);
+		tcp_sent(incomming_control_block, send_callback);
+		tcp_err(incomming_control_block, error_callback);
+		tcp_poll(incomming_control_block, poll_callback , 0);
+		return ERR_OK;
+	}else
+		return ERROR;
+}
+
+err_t ServerSocket::receive_callback(void* arg, struct tcp_pcb* client_control_block, struct pbuf* packet_buffer, err_t error){
+
+	ServerSocket* server_socket = (ServerSocket*) arg;
+	server_socket->client_control_block = client_control_block;
+
+	if(packet_buffer == nullptr){									//If RECEIVING buffer is empty, is a signal to close connection
+		server_socket->state = CLOSING;
+		pbuf_free(server_socket->tx_packet_buffer);
+		server_socket->close();
+		return ERR_OK;
+	}
+	else if(error != ERR_OK){										//Check if packet is valid
+		if(packet_buffer != nullptr){
+			pbuf_free(packet_buffer);
+		}
+		return error;
+	}
+	else if(server_socket->state == ACCEPTED){
+		if(server_socket->rx_packet_buffer == nullptr){
+			server_socket->rx_packet_buffer = packet_buffer;
+			server_socket->process_data();
+		}else{
+			pbuf_chain(server_socket->rx_packet_buffer, packet_buffer);
+		}
+		return ERR_OK;
+	}
+	else if(server_socket->state == CLOSING){
+		pbuf_free(server_socket->rx_packet_buffer);
+		server_socket->rx_packet_buffer = nullptr;
+		pbuf_free(packet_buffer);
+		return ERR_OK;
+	}
+	return ERR_OK;
+}
+
+void ServerSocket::error_callback(void *arg, err_t error){
+	ServerSocket* server_socket = (ServerSocket*) arg;
+	if(error == ERR_RST || error == ERR_ABRT){
+		server_socket->close();
+	}
+	while(true){
+		// Not all errors are covered, while True for debugging purpose
+		//TODO: Error Handler
+	}
+}
+
+err_t ServerSocket::poll_callback(void *arg, struct tcp_pcb *client_control_block){
+	ServerSocket* server_socket = (ServerSocket*)arg;
+	server_socket->client_control_block = client_control_block;
+
+	if(server_socket != nullptr){
```

Comment:
```suggestion
	if(not server_socket){
```
No te marco mas de estos, haz Ctrl+F != nullptr si lo cambias